README
NAME
    Statistics::R::IO - Perl interface to serialized R data

VERSION
    version 1.0

SYNOPSIS
        use Statistics::R::IO;
    
        my $var = Statistics::R::IO::readRDS('file.rds');
        print $var->to_pl;
    
        my %r_workspace = Statistics::R::IO::readRData('.RData');
        while (my ($var_name, $value) = each %r_workspace) {
            print $var_name, $value;
        }

        my $pi = Statistics::R::IO::evalRserve('pi');
        print $pi->to_pl;

DESCRIPTION
    This module is a pure-Perl implementation for reading native data files
    produced by the R statistical computing environment
    <http://www.r-project.org>)

    It provides routines for reading files in the two primary file formats
    used in R for serializing native objects:

    RDS RDS files store a serialization of a single R object (and, if the
        object contains references to other objects, such as environments,
        all the referenced objects as well). These files are created in R
        using the "readRDS" function and are typically named with the ".rds"
        file extension.

    RData
        RData files store a serialization of a collection of *named*
        objects, typically a workspace. These files are created in R using
        the "save" function and are typically named with the ".RData" file
        extension. (Contents of the R workspace can also be saved
        automatically on exit to the file named .RData, which is by default
        automatically read in on startup.)

    As of version 0.04, the module can also evaluate R code on a remote host
    that runs the Rserve <http://www.rforge.net/Rserve/> binary R server.
    This allows Perl programs to access all facilities of R without the need
    to have a local install of R or link to an R library.

    See "SUBROUTINES" for invocation and usage information on individual
    subroutines, and the R Internals manual
    <http://cran.r-project.org/doc/manuals/R-ints.html> for the
    specification of the file formats.

EXPORT
    Nothing by default. Optionally, subroutines "readRDS", "readRData", and
    "evalRserve", or ":all" for all three.

SUBROUTINES
    readRDS EXPR
        Reads a file in RDS format whose filename is given by EXPR and
        returns a Statistics::R::REXP object.

    readRData EXPR
        Reads a file in RData format whose filename is given by EXPR and
        returns a hash whose keys are the names of objects stored in the
        file with corresponding values as Statistics::R::REXP instances.

    evalRserve REXPR [ HOSTNAME [, PORT] | HANDLE]
        Evaluates an R expression, given as text string in REXPR, on an
        Rserve <http://www.rforge.net/Rserve/> server and returns its result
        as a Statistics::R::REXP object.

        The server location can be specified either by its host name and
        (optionally) port or by a connected instance of IO::Handle. The
        caller passing the HANDLE is responsible for reading (and checking)
        the server ID that is returned in the first 32-byte response when
        the connection was established. This allows opening the connection
        once and reusing it in multiple calls to 'evalRserve'.

        If only REXPR is given, the function assumes that the server runs on
        the localhost. If PORT is not specified, it defaults to the standard
        Rserve port, 6311.

        The function will close the connection to the Rserve host if it has
        opened it itself, but not if the connection was passed as a HANDLE.

DEPENDENCIES
    Requires perl 5.010 or newer.

  Core modules
    *   strict

    *   warnings

    *   overload

    *   Carp

    *   Exporter

    *   Module::Build

    *   Scalar::Util

    *   Test::More

  Additional CPAN modules
    *   Moose

    *   namespace::clean

    *   Test::Fatal

BUGS AND LIMITATIONS
    The module currently handles the 'version 2' serialization format, used
    since R 1.4.0 (released in December 2001). Only XDR and native-order
    binary is implemented, and since the R documentation describes the ASCII
    save format as "now mainly of historical interest", this is unlikely to
    change soon. No check is performed that a file stored in native-order
    binary was created on a platform that used the same order, and it is up
    to the caller to ensure compatibility. (Given that the default save
    format is XDR, and the prevalence of Intel platforms, this is unlikely
    to be a problem for either publicly-distributed or internal data files.)

    Data files compressed with 'gzip' and 'bzip2' are supported, but not
    'xz' ones. Again, given the R defaults ('gzip') and the fact that
    "IO::Uncompress::UnXz" is not production-ready, this is unlikely to
    change soon.

    There are some R types that are not (yet) implemented, although all
    typical "user-facing" types -- such as vectors, lists, and environments
    -- are. The remaining R types will be implemented as-needed; in other
    words, if you come across one that you need to read a particular file,
    please send me the type (the id will included in the "unimplemented
    SEXPTYPE" error message) and, if possible, how it was generated.

    There are no known bugs in this module. Please report any bugs or
    feature requests to "bug-statistics-r-io at rt.cpan.org", or through the
    web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Statistics-R-IO>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Statistics::R::IO

    You can also look for information at:

    *   RT: CPAN's request tracker (report bugs here)

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Statistics-R-IO>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Statistics-R-IO>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Statistics-R-IO>

    *   Search CPAN

        <http://search.cpan.org/dist/Statistics-R-IO/>

AUTHOR
    Davor Cubranic <cubranic@stat.ubc.ca>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2016 by University of British Columbia.

    This is free software, licensed under:

      The GNU General Public License, Version 3, June 2007



Changes
Revision history for Statistics-R-IO

1.0     28 May 2016
        - switch from Moose to Class::Tiny and Class::Tiny::Antlers (#52)
        - Implement S4 SEXPs (#49)
        - QAP parsing of tagged and untagged pairlists (#47)
        - Add a 'sexptype' attribute to REXPs
        - Add an `inherits` method to the 'REXP' class (#45)
        - Add a copy and single-argument constructor to Unknown
        - Allow specyfing image dimensions in the WebWork macro (#51)
        - Capture R error message from Rserve::eval (#43)
        - Fix failing tests wit NaN/Inf on Windows (#40)

0.101   23 December 2014
        - Fix failing tests involving comparison with infinity/NaN (#40)
        - Better handling of unopened server file handle in destructor (#41)
        - Declare testing dependency on Test::Mock::Extends
        
0.10    17 December 2014
        - Implement complex vector SEXPs (#30)
        - Implement closure SEXPs (#29, #32, #34, #35, #36, #38)
        - Implement expression SEXPs (#26)
        - Implement special environments global, base, and empty (#31,
          #33)
        - Allow configuring Rserve host from WebWork settings
        - Add graceful failure in WWk macros if Rserve host is not
          configured
        - Fix parsing of special row names (#39)

0.092   21 July 2014
        - fix setting custom Rserve host and port
        - allow specifying Rserve host for unit tests with RSERVE_HOST
          environment variable

0.091   30 June 2014
        - Lower the minimum Perl version to 5.010
        
0.08    23 June 2014
        - Use scalars in Perl rep of list for atomic vectors of length 1
        - Switch from Moo to Moose (#22)
        - Create Rserve::get_file method to copy a file from the R server
        - Allow using low-level sockets as Rserve handles
        - Include a WebWork macros that wrap the Rserve functionality

0.071   22 May 2014
        - Fix test failures when using Perl with long doubles (#9)

0.07    13 May 2014
        - Implement NA values for all vector types (#16, #17, #18)
        - Fix parsing of logical vectors (incl. #14, #4)
        - Use Perl IO for Rserve communication (#13)
        - Fix argument check to work on 32-bit Perl (#12)
        
0.06    16 April 2014
        - Fix out-of-memory crashes on OpenBSD (issue #10)
        - Fix test failures when using Perl with long doubles (#9)
        - Specify prerequisites' minimum version where necessary (#11)
        
0.05    2 April 2014
        - Provide an object-oriented API to Rserve client
        - Implement the Rserve QAP1 protocol for data serialization

0.04    27 March 2014
        Implement a basic Rserve client

0.03    21 March 2014
        Provide an object-oriented API to RDS and RData files

0.02    20 March 2014
        Fix failure on Perl 5.12 caused by using a 5.14 feature.

0.01    18 March 2014
        Initial release.



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.005.
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker;

use File::ShareDir::Install;
$File::ShareDir::Install::INCLUDE_DOTFILES = 1;
$File::ShareDir::Install::INCLUDE_DOTDIRS = 1;
install_share dist => "extras";


my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl interface to serialized R data",
  "AUTHOR" => "Davor Cubranic <cubranic\@stat.ubc.ca>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::ShareDir::Install" => "0.06"
  },
  "DISTNAME" => "Statistics-R-IO",
  "LICENSE" => "gpl",
  "MIN_PERL_VERSION" => "5.010",
  "NAME" => "Statistics::R::IO",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Class::Method::Modifiers" => 0,
    "Class::Tiny" => 0,
    "Class::Tiny::Antlers" => "0.023",
    "Math::Complex" => 0,
    "Scalar::Util" => 0,
    "namespace::clean" => "0.24"
  },
  "TEST_REQUIRES" => {
    "Test::Fatal" => 0,
    "Test::MockObject" => 0,
    "Test::MockObject::Extends" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "1.0",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Class::Method::Modifiers" => 0,
  "Class::Tiny" => 0,
  "Class::Tiny::Antlers" => "0.023",
  "Math::Complex" => 0,
  "Scalar::Util" => 0,
  "Test::Fatal" => 0,
  "Test::MockObject" => 0,
  "Test::MockObject::Extends" => 0,
  "Test::More" => "0.96",
  "namespace::clean" => "0.24"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}
